### PR TITLE
Compatibility with Codeception 5 and PHP 8.1

### PR DESCRIPTION
**Pull request by developedsoftware**
All tests pass. 

Changes to yii2-advanced-template are required and are not breaking changes. So I will submit a PR to that repo shortly - this PR is here https://github.com/yiisoft/yii2-app-advanced/pull/518

Once this and the above pull requests are accepted this module is good to go with Codecept V5 and PHP 8.1

Screenshot 2022-02-23 at 16 59 59

### CODE REVIEW

**Comment by developedsoftware** [comment]
Doesn’t look like the test suite is running 8.1? Tests fail because of this. 

Also, tests will fail until dev-master of the advanced template is updated with my changes.

Happy to keep this open until both of the above are satisfied.

File: composer.json
Comment (TavoNiievez) [inline composer.json:22]: now we can safely use `"^5.0.0-alpha2",`

Comment (TavoNiievez) [inline composer.json:22]: now we can safely use `"^5.0.0-alpha2",`

**Comment by developedsoftware** [comment]
It fails on a dependency check as another package requires dev-5.0-interfaces and it can only use one (that was the error I was getting)
________________________________

File: .gitignore
Comment (TavoNiievez) [inline .gitignore:2]: if this folder is not directly related to the project it should be part of the global .`gitignore`.
I know that in theory the same would apply to `.idea/` but it's the first time I've seen it inside `Codeception/*`

File: composer.json
Comment (TavoNiievez) [inline composer.json:28]: It's not related to Codeception 5 compatibility at all, but since this will dump a major version we could take advantage of upgrading verify to version 2.2 https://github.com/Codeception/Verify/blob/master/UPGRADE.md

File: composer.json
Comment (TavoNiievez) [inline composer.json:21]: Is there a compelling reason not to allow `PHP 8.0`?

**Review by TavoNiievez — COMMENTED** [review]
Excellent work!

File: composer.json
Comment (TavoNiievez) [reply composer.json:22]: @developedsoftware You have to identify the module that is causing the conflict and update it there too `"^5.0.0-alpha2"` is the correct restriction.

**Comment by developedsoftware** [comment]
Ok no problem. I will action all the above points shortly.

File: composer.json
Comment (developedsoftware) [reply composer.json:21]: No my mistake. I thought Codeception v5.0 required 8.1 but I can see it only requires 8.

File: composer.json
Comment (developedsoftware) [reply composer.json:22]: OK done -> https://github.com/Codeception/module-filesystem/pull/18

File: .gitignore
Comment (developedsoftware) [reply .gitignore:2]: OK removed.

File: composer.json
Comment (developedsoftware) [reply composer.json:28]:
```
Problem 2
    - Root composer.json requires codeception/codeception ^5.0.0-alpha2 -> satisfiable by codeception/codeception[5.0.0-alpha2, 5.0.x-dev].
    - codeception/verify 2.2.0 requires phpunit/phpunit ^9.3 -> satisfiable by phpunit/phpunit[9.3.0, ..., 9.5.x-dev].
    - You can only install one version of a package, so only one of these can be installed: phpunit/phpunit[dev-master, 7.5.0, ..., 7.5.20, 8.0.0, ..., 8.5.x-dev, 9.0.0, ..., 9.5.x-dev].
    - codeception/codeception[5.0.0-alpha2, ..., 5.0.x-dev] require phpunit/phpunit dev-master as 9.5.99 -> satisfiable by phpunit/phpunit[dev-master].
    - Root composer.json requires codeception/verify ^2.2 -> satisfiable by codeception/verify[2.2.0].

```

https://github.com/Codeception/Verify/pull/53

**Comment by developedsoftware** [comment]
Changes made but blocked by other dependencies. 

Your requirements could not be resolved to an installable set of packages.

```
  Problem 1
    - Root composer.json requires codeception/module-filesystem ^3.0 -> satisfiable by codeception/module-filesystem[3.0.x-dev].
    - codeception/module-filesystem 3.0.x-dev requires codeception/codeception dev-5.0-interfaces as 5.0.0 -> found codeception/codeception[dev-5.0-interfaces] but it conflicts with your root composer.json require (^5.0.0-alpha2).
  Problem 2
    - Root composer.json requires codeception/codeception ^5.0.0-alpha2 -> satisfiable by codeception/codeception[5.0.0-alpha2, 5.0.x-dev].
    - codeception/verify dev-master requires phpunit/phpunit ^9.3 -> satisfiable by phpunit/phpunit[9.3.0, ..., 9.5.x-dev].
    - You can only install one version of a package, so only one of these can be installed: phpunit/phpunit[dev-master, 7.5.0, ..., 7.5.20, 8.0.0, ..., 8.5.x-dev, 9.0.0, ..., 9.5.x-dev].
    - codeception/codeception[5.0.0-alpha2, ..., 5.0.x-dev] require phpunit/phpunit dev-master as 9.5.99 -> satisfiable by phpunit/phpunit[dev-master].
    - Root composer.json requires codeception/verify dev-master -> satisfiable by codeception/verify[dev-master].
```

Will keep this open until the above can be satisfied. Cheers!

File: tests/bootstrap.php
Comment (samdark) [inline tests/bootstrap.php:5]: What is the reason for adding the closure here? I see no scope to guard.

Comment (developedsoftware) [reply tests/bootstrap.php:5]: I copied the bootstrap file inside yii advanced template (was getting errors that Yii was undefined)

Comment (developedsoftware) [reply tests/bootstrap.php]: I have reverted back. The reason I was getting errors was because vendor/autoload.php was not loaded so I am including this also. Latest push essentially just removes the closure as you rightly noticed.

**Review by samdark — APPROVED** [review]

File: composer.json
Comment (TavoNiievez) [reply composer.json]: @developedsoftware you're ready to go now!

**Comment by developedsoftware** [comment]
I will fix remaining tests now I can get composer running on my local machine.

**Comment by developedsoftware** [comment]
I am getting errors that I was not before inside the yii2-advanced-template. Am I missing something really obvious?

```
frontend\tests\unit\models\ContactFormTest::testSendEmail()
Error: Call to undefined function frontend\tests\unit\models\expect_that()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/ContactFormTest.php:21
/usr/local/bin/codecept:5

frontend\tests\unit\models\PasswordResetRequestFormTest::testSendMessageWithWrongEmailAddress()
Error: Call to undefined function frontend\tests\unit\models\expect_not()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/PasswordResetRequestFormTest.php:32
/usr/local/bin/codecept:5

frontend\tests\unit\models\PasswordResetRequestFormTest::testNotSendEmailsToInactiveUser()
Error: Call to undefined function frontend\tests\unit\models\expect_not()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/PasswordResetRequestFormTest.php:40
/usr/local/bin/codecept:5

frontend\tests\unit\models\PasswordResetRequestFormTest::testSendEmailSuccessfully()
Error: Call to undefined function frontend\tests\unit\models\expect_that()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/PasswordResetRequestFormTest.php:51
/usr/local/bin/codecept:5

frontend\tests\unit\models\ResendVerificationEmailFormTest::testWrongEmailAddress()
Error: Call to undefined method Codeception\Verify\Expectations\ExpectAny::false()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/ResendVerificationEmailFormTest.php:35
/usr/local/bin/codecept:5

frontend\tests\unit\models\ResendVerificationEmailFormTest::testEmptyEmailAddress()
Error: Call to undefined method Codeception\Verify\Expectations\ExpectAny::false()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/ResendVerificationEmailFormTest.php:47
/usr/local/bin/codecept:5

frontend\tests\unit\models\ResendVerificationEmailFormTest::testResendToActiveUser()
Error: Call to undefined method Codeception\Verify\Expectations\ExpectAny::false()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/ResendVerificationEmailFormTest.php:59
/usr/local/bin/codecept:5

frontend\tests\unit\models\ResendVerificationEmailFormTest::testSuccessfullyResend()
Error: Call to undefined method Codeception\Verify\Expectations\ExpectAny::true()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/ResendVerificationEmailFormTest.php:71
/usr/local/bin/codecept:5

frontend\tests\unit\models\ResetPasswordFormTest::testResetCorrectToken()
Error: Call to undefined function frontend\tests\unit\models\expect_that()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/ResetPasswordFormTest.php:41
/usr/local/bin/codecept:5

frontend\tests\unit\models\SignupFormTest::testCorrectSignup()
Error: Call to undefined method Codeception\Verify\Expectations\ExpectAny::true()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/SignupFormTest.php:35
/usr/local/bin/codecept:5

frontend\tests\unit\models\SignupFormTest::testNotCorrectSignup()
Error: Call to undefined function frontend\tests\unit\models\expect_not()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/SignupFormTest.php:63
/usr/local/bin/codecept:5

frontend\tests\unit\models\VerifyEmailFormTest::testVerifyCorrectToken()
Error: Call to undefined method Codeception\Verify\Expectations\ExpectAny::isInstanceOf()
/Users/luke.english/NetBeansProjects/codeception/module-yii2/vendor/yiisoft/yii2-app-advanced/frontend/tests/unit/models/VerifyEmailFormTest.php:48
/usr/local/bin/codecept:5

Full output can be found in Output window.
Perhaps error occurred, verify in Output window.
```

**Comment by developedsoftware** [comment]
I dont think I can upgrade codeception/verify to 2.2

It breaks the tests in yii2 which is able to run on php 5!

https://github.com/yiisoft/yii2-app-advanced/pull/519

Propose dropping back down to "codeception/verify": "<2"

Thoughts?

**Comment by developedsoftware** [comment]
With codeception/verify dropped back down to the original version, all tests now pass...

File: composer.json
Comment (developedsoftware) [reply composer.json:28]: I dont think I can upgrade codeception/verify to 2.2

It breaks the tests in yii2 which is able to run on php 5!

https://github.com/yiisoft/yii2-app-advanced/pull/519

Propose dropping back down to "codeception/verify": "<2"

Thoughts?

**Comment by TavoNiievez** [comment]
@developedsoftware https://github.com/yiisoft/yii2-app-advanced/pull/520

File: composer.json
Comment (TavoNiievez) [reply composer.json:28]: gh doesn't notify me if you edit a comment, make sure to tag me when you do.

Comment (developedsoftware) [reply composer.json:28]: Ok I think this PR is good to go. But I agree verify 2.2 is worth upgrading to if yii2 is able to drop php5 and require a min version of php7

Comment (developedsoftware) [reply composer.json:28]: Ok I think this PR is good to go. But I agree verify 2.2 is worth upgrading to if yii2 is able to drop php5 and require a min version of php7

**Comment by Naktibalda** [comment]
I was confident that I released codeception/module-filesystem 3.0.0, but apparently I didn't.

master branch was codeception 5 ready, we have to decide if we keep master branch or 3.0 branch.

**Comment by developedsoftware** [comment]
I think a v3.0 branch would be good to keep it consistent with the other repos. But ultimately it doesn’t matter right ?

**Comment by Naktibalda** [comment]
Master branch is more consistent.

Most modules have master branch only, in some modules 1.x or 1.0 braches were created for applying patches to older versions.

Getting rid of master branch would be a new thing.

**Comment by developedsoftware** [comment]
Maybe keep it in master branch and just tag the relevant release if it has breaking changes ?

**Review by TavoNiievez — APPROVED** [review]

File: composer.json
Comment (developedsoftware) [reply composer.json:28]: > It's not related to Codeception 5 compatibility at all, but since this will dump a major version we could take advantage of upgrading verify to version 2.2 https://github.com/Codeception/Verify/blob/master/UPGRADE.md

When pull request https://github.com/yiisoft/yii2-app-advanced/pull/521 is merged we should be able to bump to codeception/verify 2.2

Comment (developedsoftware) [reply composer.json]: Tests now pass.
Screenshot 2022-02-25 at 09 11 01

**Comment by developedsoftware** [comment]
Codecept/verify 2.2 working and all tests pass. Hoping this can now be merged in once approved by everyone?

**Review by samdark — APPROVED** [review]

**Review by TavoNiievez — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**
